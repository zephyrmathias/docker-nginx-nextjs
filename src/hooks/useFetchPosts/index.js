import { useState, useEffect } from 'react'
import axios from 'axios'

function useFetchPosts() {
  const [posts, setPosts] = useState([])
  const [error, setError] = useState(null)

  useEffect(() => {
    let ignore = false

    async function fetchPosts() {
      try {
        setError(null)
        const { data } = await axios.get(
          'https://jsonplaceholder.typicode.com/posts',
        )
        if (!ignore) setPosts(data)
      } catch (err) {
        if (!ignore) setError(err)
      }
    }

    fetchPosts()

    return () => {
      ignore = true
    }
  }, [])

  return { posts, error }
}

export default useFetchPosts
