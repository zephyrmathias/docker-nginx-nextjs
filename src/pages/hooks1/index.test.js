import React from 'react'
import axios from 'axios'
import { render, act, waitForDomChange } from '@testing-library/react'
import useFetchPosts from '../../hooks/useFetchPosts'
import CustomHooks1 from './index'

jest.mock('../../hooks/useFetchPosts')

// unit test with mocking useFetchPosts directly
test('should match with snapshot when it contains a list of posts', () => {
  useFetchPosts.mockImplementation(() => ({
    posts: [
      {
        userId: 1,
        id: 1,
        title: 'mock title',
        body: 'mock body',
      },
    ],
    error: null,
  }))

  act(() => {
    const { asFragment } = render(<CustomHooks1 />)
    expect(asFragment()).toMatchSnapshot()
  })
})

// this is integration testing but mocking axios instead
// test('should match with snapshot when it contains a list of posts', async () => {
//   const mockResult = [
//     {
//       userId: 1,
//       id: 1,
//       title: 'mock title',
//       body: 'mock body',
//     },
//   ]
//   axios.get.mockResolvedValue({ data: mockResult })

//   await act(async () => {
//     const { asFragment } = render(<CustomHooks1 />)
//     await waitForDomChange()
//     expect(asFragment()).toMatchSnapshot()
//   })
// })

test('should match with snapshot when it is error', () => {
  useFetchPosts.mockImplementation(() => ({
    posts: [],
    error: 'error message',
  }))

  act(() => {
    const { asFragment } = render(<CustomHooks1 />)
    expect(asFragment()).toMatchSnapshot()
  })
})

test('should match with snapshot when it does not contain a post', () => {
  useFetchPosts.mockImplementation(() => ({
    posts: [],
    error: null,
  }))

  act(() => {
    const { asFragment } = render(<CustomHooks1 />)
    expect(asFragment()).toMatchSnapshot()
  })
})
