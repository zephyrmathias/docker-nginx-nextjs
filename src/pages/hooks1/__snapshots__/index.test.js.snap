// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match with snapshot when it contains a list of posts 1`] = `
<DocumentFragment>
  <div>
    <div>
      This is the example of Custom Hooks (API Fetch) with UseEffect
    </div>
    <div>
      <h4>
        mock title
      </h4>
      <p>
        mock body
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match with snapshot when it does not contain a post 1`] = `
<DocumentFragment>
  <div>
    Loading..
  </div>
</DocumentFragment>
`;

exports[`should match with snapshot when it is error 1`] = `
<DocumentFragment>
  <div>
    Error!
  </div>
</DocumentFragment>
`;
